lambda/matching.ml

```diff
@@ -2819,48 +2880,92 @@ let combine_constructor loc arg pat_env cstr partial ctx def
-      let consts, nonconsts =
-        split_cases (List.map tag_lambda descr_lambda_list) in
       let lambda1 =
         match (fail_opt, same_actions descr_lambda_list) with
         | None, Some act -> act (* Identical actions, no failure *)
         | _ -> (
+            let cases = split_cases (List.map tag_lambda descr_lambda_list) in
             match
-              (cstr.cstr_consts, cstr.cstr_nonconsts, consts, nonconsts)
+              (cstr.cstr_consts, cstr.cstr_nonconsts, cases)
             with
-            | 1, 1, [ (0, act1) ], [ (0, act2) ] ->
+            | 1, 1, { consts = [ (0, act1) ]; nonconsts = [ (0, act2) ] }
                 (* Typically, match on lists, will avoid isint primitive in that
               case *)
+            | 1, 0, { consts = [ (0, act1) ]; any_nonconst = Some act2 } ->
                 Lifthenelse (arg, act2, act1)
-            | n, 0, _, [] ->
+            | n, 0, { any_const; nonconsts = []; any_nonconst = None } ->
                 (* The type defines constant constructors only *)
-                call_switcher loc fail_opt arg 0 (n - 1) consts
-            | n, _, _, _ -> (
-                let act0 =
-                  (* = Some act when all non-const constructors match to act *)
-                  match (fail_opt, nonconsts) with
-                  | Some a, [] -> Some a
-                  | Some _, _ ->
-                      if List.length nonconsts = cstr.cstr_nonconsts then
-                        same_actions nonconsts
+               begin match any_const with
+               | Some act -> act
+               | None ->
+                  call_switcher loc fail_opt arg 0 (n - 1) cases.consts
+               end
+            | n, _, _ -> (
+                let single_action any_case cases complete_case_count =
+                  (* = Some act when all actions are the same,
+                     either on constant or non-constant domain *)
+                  match (fail_opt, any_case, cases) with
+                  | Some a, None, []
+                  | _, Some a, _ -> Some a
+                  | None, None, nonconsts -> same_actions nonconsts
+                  | Some _, None, cases ->
+                    (* FIXME: today cstr.cstr_{non,}consts does not count
+                       unboxed constructors so it cannot be correctly used
+                       to decide exhaustiveness.
+
+                       TODO: have a version of cstr_nonconsts that counts
+                       not the number of different constructors, but the number
+                       of possible head tags (in the non-any case); then we can
+                       in fact compare this number to the length of the list
+                       -- given that split_cases will expand several-tag
+                       constructors into several list items. *)
+                      if List.length cases = complete_case_count then
+                        same_actions cases
```

Comment:
As far as I can tell, this `FIXME` is left over in the complete PR, and this `TODO` is not handled (and, if we wanted to follow its suggestion, would require adding yet another field to the Datarepr stuff). What is your plan for those two?